index.php

```diff
@@ -28,51 +28,23 @@ function error_page($header, $body)
     error_page('Configuration Error', 'Endpoint not yet configured, visit <a href="setup.php">setup.php</a> for instructions on how to set it up.');
 }
 
-//temp code generation to protect against CSRF, this is only going to be valid for between 2 and 4 minutes
-function generate_csrf_code($redirect_uri, $client_id, $state)
+function create_signed_code($key, $message, $ttl = 31536000, $appended_data = '')
 {
-    $t = time();
-    $m = intval(date('i', $t) / 2) * 2;
-    return md5(APP_KEY . $client_id . $redirect_uri . $state . date('Y-M-d G:', $t) . $m);
-
-}
-
-function verify_csrf_code($redirect_uri, $client_id, $state, $code)
-{
-    $t = time();
-    $t2 = time() - 120;
-    $m = intval(date('i', $t) / 2) * 2;
-    $m2 = intval(date('i', $t2) / 2) * 2;
-    if( md5(APP_KEY . $client_id . $redirect_uri . $state .  date('Y-M-d G:', $t) . $m) == $code 
-            || md5(APP_KEY . $client_id . $redirect_uri . $state . date('Y-M-d G:', $t2) . $m2) == $code ) {
-        return true;
-    }
-    return false;
+    $expires = time() + $ttl;
+    $body = $message . $expires . $appended_data;
+    $signature = hash_hmac('sha256', $body, $key);
+    return dechex($expires) . ':' . $signature . ':' . $appended_data;
 }
 
-
-//temp code generation, this is only going to be valid for between 5 and 10 minutes
-function generate_code($redirect_uri, $client_id, $scope)
+function verify_signed_code($key, $message, $code)
 {
-    $t = time();
-    $m = intval(date('i', $t) / 5) * 5;
-    return md5(APP_KEY . USER_URL . $redirect_uri . $client_id . $scope . date('Y-M-d G:', $t) . $m);
-
-}
-
-function verify_code($redirect_uri, $client_id, $scope, $code)
-{
-    $t = time();
-    $t2 = time() - 300;
-    $m = intval(date('i', $t) / 5) * 5;
-    $m2 = intval(date('i', $t2) / 5) * 5;
-
-    if( md5(APP_KEY . USER_URL . $redirect_uri . $client_id . $scope . date('Y-M-d G:', $t) . $m) == $code 
-            || md5(APP_KEY . USER_URL . $redirect_uri . $client_id . $scope . date('Y-M-d G:', $t2) . $m2) == $code ) {
-        return true;
-
-    }
-    return false;
+    $code_parts = explode(':', $code, 3);
+    if (count($code_parts) !== 3) return false;
+    $expires = hexdec($code_parts[0]);
+    if (time() > $expires) return false;
+    $body = $message . $expires . $code_parts[2];
+    $signature = hash_hmac('sha256', $body, $key);
+    return $signature === $code_parts[1];
```

Comment:
as discussed in IRC, would prefer to check for hash_equals and use that, thus we only leave < 5.6 open to timing attacks